### ISSUE TITLE
Decide how to do BigQuery testing for the hspec test suite

### ISSUE THREAD

**Issue by chrisdone** [hasura/graphql-engine#7929]
BigQuery is currently the only backend that will not setup a fresh schema for each tested feature. MySQL, PostgreSQL, SQL Server [all do](https://github.com/hasura/graphql-engine-mono/blob/63bdc67ac7ee23e798887efae310bfff9c02070e/server/tests-hspec/BasicFieldsSpec.hs#L28-L41).

It would be cleaner and more code-driven to have the test suite, and further, the individual feature test, dictate what the schema is and what the data is inside.

Options:

1. **Status Quo:** Keep existing approach, where BigQuery's schema and data are set in stone in the cloud. 
1. **Validate:** Have a validation/sanity check for each feature that tests that the schema in the database matches exactly the schema that we expect.
1.  **Fresh projects**: Use the Google APIs to create projects, but requires an "Organization" to be able to make projects fresh via the API. Each temporary project would have a dataset named "hasura" created fresh for that test run.
1. **Global Mutex**: Have a dataset locking mechanism: e.g. make a `hasura` dataset and a `lock` dataset, and make a mutex out of the lock dataset so that only one user/CI can work on `hasura` at once, thereby permitting the use of setup/teardowns.
1.  **Randomized names**: Change the name of the schema to be generated per run, but then the other backends would need to be similarly clever: currently, they all assume the schema is called `hasura` and [that is created by docker-compose](https://github.com/hasura/graphql-engine-mono/blob/63bdc67ac7ee23e798887efae310bfff9c02070e/docker-compose.yml#L32).
1. **Isolated accounts**: Have everybody each have their own BigQuery account and CI will implement a mutex for its own purposes.

Pros and cons to each.

Relates to https://github.com/hasura/graphql-engine/issues/7755

**Comment by sassela** [comment]
small update on option 3: we already have a GCP Organization set up and I can create new projects via the API from my own Hasura account. But I need additional permissions to create and configure a separate [resource folder](https://cloud.google.com/resource-manager), within which to contain ephemeral projects. I'll update here when I have them.

**Comment by chrisdone** [comment]
Tip top news! Looking forward to updates.

**Comment by sassela** [comment]
FYI, I'm now trying out ephemeral BigQuery projects for the pytest suite. Will let you know once I have a small working example we can use for the hspec test suite!

**Comment by chrisdone** [comment]
That sounds good.

**Comment by chrisdone** [comment]
I assume we're waiting for someone else from the org on this?

**Comment by sassela** [comment]
no, just me. here's [my WIP](https://github.com/hasura/graphql-engine-mono/pull/3240/files) to create and use a new project per testrun in CI. Making slow progress as I'm context switching. Let's talk tomorrow if you're blocked; BigQuery testing in hspec doesn't need to be reliant on me trying this for the pytests.

**Comment by qrilka** [comment]
During our last call with @sassela and @aniketd it was decided that for now we'll keep the current approach for using BigQuery in python tests - i.e. rely on existence of service account and project id as environment variables.
The original idea to come up with something like what was done for PG and other DBs using docker-compose isn't seem to be applicable to BigQuery because:
* it's an external service and thus it's not possible to spawn a new server with hardcoded auth details
* the code in https://github.com/hasura/graphql-engine-mono/pull/3240 for creating projects dynamically is aimed at CI, it relies on having a Google Cloud org, it's not needed for local runs and such a setup is too complex for an individual contributor

**Comment by sassela** [comment]
thanks for the summary @qrilka ! To clarify, whilst we can use the existing setup for local tests, I _would_ like us to use this approach (option 3, short-lived projects) when running the hspec tests, including those for BigQuery, in CI.

**Comment by qrilka** [comment]
The implementation in https://github.com/hasura/graphql-engine-mono/pull/3435 has currently failing tests because of the BigQuery backend returning numbers as strings. There is a ticket https://hasurahelp.zendesk.com/hc/en-us/requests/1551 on this topic which should unblock this

**Comment by sassela** [comment]
Thanks @qrilka! I’m still a little unsure how to proceed tbh. @0x777, do you have any advice?

Restating some of the [original goals](https://github.com/hasura/graphql-engine/blob/master/rfcs/hspec-test-suite.md#project-goal) of this suite I think are relevant:
1. Make it as easy as possible for contributors to understand, add and edit tests for features across any / all backends
2. Minimise duplication of setup, teardown and queries for feature sets that are common across backends

Another implicit (i.e. not mentioned in [the RFC](https://github.com/hasura/graphql-engine/blob/master/rfcs/hspec-test-suite.md)) ideal that I see is that we should only test the behaviour as it exists in the graphql-engine, rather than introduce or await workarounds for easier testing.

Given that, options I can think of:
1. Block this set of tests on [this request to recast strings](https://hasurahelp.zendesk.com/agent/tickets/1551) in the graphql-engine. I’m against this as I’d prefer to explicitly test the existing behaviour, as above. Also, it’s not clear to me from the linked ticked discussion whether this will be worked on at all.
2. Recast strings as numerics _in test only_, or compare stringified `id`s in `shouldReturnYaml`. Again, uncomfortable with departing from the production behaviour in test for the same reason, but maybe this is the least bad approach?
3. Keep BQ tests separate, almost entirely duplicated, except the id column type. [This is what we currently do in the pytests](https://github.com/hasura/graphql-engine/blob/master/server/tests-py/test_graphql_queries.py#L116). Goes against goal 2 above, at least.
4. Unify `id` column types across all backends in spec (i.e. serialise or stringify all IDs in test setup). This seems like a huge workaround, affecting all tests, for the sake of atypical BigQuery behaviour.
5. Exclude `id` from expected responses. Ditto.

I also realised I'm making the assumption that most of the tests are failing on the `id` type mismatch. Let me know if that's incorrect, or if I'm missing anything else?

**Comment by qrilka** [comment]
@sassela I think you're quite right with one extra thought from my side - currently it's only `id` in the tests that we have but we'll get into the same situation with e.g. aggregation functions and probably other places where numeric values are needed.
As for the options available I also think that the option 2 looks bad but the best at the moment. I will take a look into if it's possible to do that without polluting test code with extra details for BigQuery only.

**Comment by qrilka** [comment]
@sassela I've added `BackendOptions` and pass it explicitly to `shouldReturnYaml` in tests (see https://github.com/hasura/graphql-engine-mono/pull/3435/commits/dc79f556d1ceaafa093512a4a8b7ea64a5e5e15d). Could you take a look and share your opinion?

**Comment by sassela** [comment]
@qrilka it is an invasive change but I can’t think of a better alternative. So let’s go with this to unblock the PR 👍🏽

**Comment by sassela** [comment]
addressed by https://github.com/hasura/graphql-engine/commit/773870f44344f94dd7c4ae8996b41b9be821b3bc

**Comment by sassela** [comment]
> Recast strings as numerics in test only, or compare stringified ids in shouldReturnYaml. Again, uncomfortable with departing from the production behaviour in test for the same reason, but maybe this is the least bad approach?

The decision to go with this option caused a bit of confusion recently. I've made [a note](https://github.com/hasura/graphql-engine-mono/issues/4294) to document this better where we can.
